File: stacks/minMaxStack/minMaxStack.py
#  O(1) time | O(n) space
class MinMaxStack:
    def __init__(self):
        self.minMaxStack = []
        self.stack = []

    def peek(self):
        return self.stack[len(self.stack) - 1]

    def pop(self):
        self.minMaxStack.pop()
        return self.stack.pop()

    def push(self, number):
        newMinMax = {"min": number, "max": number}
        if len(self.minMaxStack):
            currentMinMax = self.minMaxStack[len(self.minMaxStack) - 1]
            newMinMax["min"] = min(currentMinMax["min"], number)
            newMinMax["max"] = max(currentMinMax["max"], number)

        self.stack.append(number)
        self.minMaxStack.append(newMinMax)

    def getMin(self):
        return self.minMaxStack[len(self.minMaxStack) - 1]["min"]

    def getMax(self):
        return self.minMaxStack[len(self.minMaxStack) - 1]["max"]


# Example usage:
if __name__ == "__main__":
    stack = MinMaxStack()
    stack.push(5)
    stack.push(3)
    stack.push(7)
    print("Min:", stack.getMin())  # Output: Min: 3
    print("Max:", stack.getMax())  # Output: Max: 7
    print("Peek:", stack.peek())  # Output: Peek: 7
    print("Pop:", stack.pop())  # Output: Pop: 7
    print("Min after pop:", stack.getMin())  # Output: Min after pop: 3
    print("Max after pop:", stack.getMax())  # Output: Max after pop: 5
    print("Peek after pop:", stack.peek())  # Output: Peek after pop: 3

File: stacks/minMaxStack/minMaxStack.js
// Implemention of a new stack that supports push, pop, peek, getMin, and getMax
// O(n) space complexity 

class MinMaxStack {
  arr = []
  min = []
  max = []
  
  peek() {
    const peekIndex = this.arr.length -1
    return this.arr[peekIndex]
  }

  pop() {
    const removedNumber = this.arr.pop()
    
    this.max.pop()
    this.min.pop()
    
    return removedNumber
  }

  push(number) {
    this.arr.push(number)
    if(this.max.length === 0 || number > this.max[this.max.length - 1]) {
      this.max.push(number)
    } else {
      this.max.push(this.max[this.max.length - 1])
    }
    if(this.min.length === 0 || number < this.min[this.min.length -1]) {
      this.min.push(number)
    } else {
      this.min.push(this.min[this.min.length - 1])
    }
  }

  getMin() {
    const minIndex = this.min.length -1
    return this.min[minIndex]

  }

  getMax() {
    const maxIndex = this.max.length -1
    return this.max[maxIndex]
  }
}

// Example usage
const stack = new MinMaxStack();
stack.push(5);
stack.push(3);      
stack.push(7);
console.log(stack.peek()); // 7
console.log(stack.getMin()); // 3
console.log(stack.getMax()); // 7
stack.pop();
console.log(stack.peek()); // 3
console.log(stack.getMin()); // 3
console.log(stack.getMax()); // 5